evm/src/cpu/kernel/asm/mpt/delete/delete_branch.asm

```diff
@@ -0,0 +1,151 @@
+update_branch:
+    // stack: updated_child_ptr, first_nibble, node_payload_ptr, retdest
+    %get_trie_data_size
+    // stack: updated_branch_ptr, updated_child_ptr, first_nibble, node_payload_ptr, retdest
+    PUSH @MPT_NODE_BRANCH %append_to_trie_data
+    // stack: updated_branch_ptr, updated_child_ptr, first_nibble, node_payload_ptr, retdest
+    DUP4                %mload_trie_data %append_to_trie_data // Copy child[0]
+    DUP4 %add_const(1)  %mload_trie_data %append_to_trie_data // ...
+    DUP4 %add_const(2)  %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(3)  %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(4)  %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(5)  %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(6)  %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(7)  %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(8)  %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(9)  %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(10) %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(11) %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(12) %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(13) %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(14) %mload_trie_data %append_to_trie_data
+    DUP4 %add_const(15) %mload_trie_data %append_to_trie_data // Copy child[15]
+    DUP4 %add_const(16) %mload_trie_data %append_to_trie_data // Copy value_ptr
+    // stack: updated_branch_ptr, updated_child_ptr, first_nibble, node_payload_ptr, retdest
+    SWAP1
+    // stack: updated_child_ptr, updated_branch_ptr, first_nibble, node_payload_ptr, retdest
+    DUP2 %increment DUP4 ADD
+    // stack: updated_branch_ptr+first_nibble, updated_child_ptr, updated_branch_ptr, first_nibble, node_payload_ptr, retdest
+    %mstore_trie_data
+    %stack (updated_branch_ptr, first_nibble, node_payload_ptr, retdest) -> (retdest, updated_branch_ptr)
+    JUMP
+
+// The updated child is empty. Count how many non-empty children the branch node has.
+// If it's one, transform the branch node into an leaf/extension node and return it.
+maybe_normalize_branch:
+    // stack: updated_child_ptr, first_nibble, node_payload_ptr, retdest
+    PUSH 0 %mstore_kernel_general(0) PUSH 0 %mstore_kernel_general(1)
+    // stack: updated_child_ptr, first_nibble, node_payload_ptr, retdest
+    PUSH 0
+// Loop from i=0..16 excluding `first_nibble` and store the number of non-empty children in
+// KernelGeneral[0]. Also store the last non-empty child in KernelGeneral[1].
+loop:
+    // stack: i, updated_child_ptr, first_nibble, node_payload_ptr, retdest
+    DUP1 DUP4 EQ %jumpi(loop_eq_first_nibble)
+    // stack: i, updated_child_ptr, first_nibble, node_payload_ptr, retdest
+    DUP1 %eq_const(16) %jumpi(loop_end)
+    DUP1 DUP5 ADD %mload_trie_data %mload_trie_data ISZERO ISZERO %jumpi(loop_non_empty)
```

Comment:
Second `%mload_trie_data` seems redundant: if we have an empty node, it should already be 0.

Reply:
This is in case we have an empty node not at the 0 location in the `TrieData`. I don't think it can happen currently, but I think it's better to keep the option open.